# Usando For em Listas



Nas aulas passadas aprendemos a usar o for usando a função range

In [7]:
for i in range(5):
    print(f"Index atual: {i}")

Index atual: 0
Index atual: 1
Index atual: 2
Index atual: 3
Index atual: 4


### Hoje aprenderemos a usar a estrutura de repetição For em Listas!

```
for item in lista:
    bloco de comandos
```


In [2]:
produtos = ["Notebook", "Monitor", "Teclado", "Mouse"]

for item in produtos:
    print(item)

Notebook
Monitor
Teclado
Mouse


In [4]:
# Vamos usar o for para processar uma lista de valores, algo muito comum em sistemas de e-commerce
prices = [100.00, 50.00, 20.00]
total = 0

for value in prices:
    total = total + value

print(f"O total da compra é R$ {total}")

O total da compra é R$ 170.0


A variável total começa em 0.

O for pega o 1º preço (100) e soma ao total (0 + 100 = 100).

O for pega o 2º preço (50) e soma ao total (100 + 50 = 150).

O for pega o 3º preço (20) e soma ao total (150 + 20 = 170).

In [6]:
# Cenário: Você tem uma lista de faturamentos e precisa criar uma nova lista contendo apenas as "Vendas Grandes" (acima de R$ 500,00).
#Vamos aprender a usar o if dentro do for
vendas = [120, 800, 450, 1200, 30, 600]
vendas_grandes = []

for valor in vendas:
    if valor > 500:
        vendas_grandes.append(valor)

print(f"Vendas acima de 500: {vendas_grandes}")
# Saída: [800, 1200, 600]

Vendas acima de 500: [800, 1200, 600]


In [9]:
#Cenário: O gerente da loja decidiu dar 10% de desconto em todos os produtos para a Black Friday. Vamos gerar uma nova lista com os preços atualizados.
precos_originais = [100, 250, 40, 500]
precos_com_desconto = []

for preco in precos_originais:
    novo_preco = preco * 0.90 # Tira 10%
    precos_com_desconto.append(novo_preco)

print(f"Preços promocionais: {precos_com_desconto}")
# Saída: [90.0, 225.0, 36.0, 450.0]

Preços promocionais: [90.0, 225.0, 36.0, 450.0]


In [11]:
# Entendendo o uso do break ao percorrer listas
#Cenário: Você está procurando por um produto específico em um estoque gigante. Assim que encontrar o primeiro, você quer parar de procurar para economizar processamento.

estoque = ["Parafuso", "Prego", "Martelo", "Prego", "Alicate"]
item_procurado = "Martelo"

for item in estoque:
    print(f"Verificando: {item}...")
    if item == item_procurado:
        print(f"Encontrei o {item_procurado}! Parando a busca.")
        break # O loop encerra imediatamente aqui

Verificando: Parafuso...
Verificando: Prego...
Verificando: Martelo...
Encontrei o Martelo! Parando a busca.


### Desafio: Calculando o Bônus
10min

Cenário: Uma equipe de vendas bateu a meta. Cada vendedor receberá R$ 100 de bônus sobre o seu valor de venda atual.

Lista: vendas_mes = [1500, 2300, 1200, 5000]

Objetivo: Criar uma nova lista chamada vendas_com_bonus onde cada valor seja o original mais o bônus.

In [12]:
vendas_mes = [1500, 2300, 1200, 5000]
vendas_com_bonus = []

for valor in vendas_mes:
    novo_valor = valor + 100
    vendas_com_bonus.append(novo_valor)

print(vendas_com_bonus)

[1600, 2400, 1300, 5100]


**Tratamento de dados**

Enunciado: Normalização de Dados Cadastrais (CEP)

Cenário: Durante a migração de um banco de dados legado, você recebeu uma lista de entradas de endereços onde o campo "CEP" não possui validação. Você precisa criar um script de tratamento de dados para garantir a integridade dos dados antes da inserção no novo sistema.

In [13]:
# Dados brutos recebidos da fonte (raw data)
lista_ceps_entrada = [
    "01001-000",
    " 20040001",
    "70040-010 ",
    "80010.000",
    "04571010"
]

# Lista para armazenar os dados processados e validados
lista_ceps_normalizados = []

for registro_cep in lista_ceps_entrada:
    # Aplicando a cadeia de métodos para limpeza de caracteres não numéricos
    cep_formatado = registro_cep.strip().replace("-", "").replace(".", "")

    # Armazenando o valor tratado
    lista_ceps_normalizados.append(cep_formatado)

    print(f"Processado: {registro_cep} -> {cep_formatado}")

print(lista_ceps_normalizados)

Processado: 01001-000 -> 01001000
Processado:  20040001 -> 20040001
Processado: 70040-010  -> 70040010
Processado: 80010.000 -> 80010000
Processado: 04571010 -> 04571010
['01001000', '20040001', '70040010', '80010000', '04571010']


In [14]:
# Avançando em um cenário mais complexo! Garantindo a checagem de todas as entradas de dados

lista_ceps_entrada = ["01001-000", "20040", "70040-010 ", "CEP-INVALIDO", "04571010"]

lista_ceps_validados = []
lista_ceps_rejeitados = []

for registro in lista_ceps_entrada:
    # 1. Limpeza
    cep_limpo = registro.strip().replace("-", "").replace(".", "")

    # 2. Validação de Regra de Negócio (Exatamente 8 caracteres e apenas números)
    if len(cep_limpo) == 8 and cep_limpo.isdigit():
        lista_ceps_validados.append(cep_limpo)
    else:
        lista_ceps_rejeitados.append(registro)

print(f"Sucesso: {lista_ceps_validados}")
print(f"Rejeitados: {lista_ceps_rejeitados}")

Sucesso: ['01001000', '70040010', '04571010']
Rejeitados: ['20040', 'CEP-INVALIDO']


### Lista de exercícios adicionais

**Ex1: Tratamento de dados de produtos**

Cenário: Você recebeu uma lista de códigos de produtos (SKUs) de um fornecedor. Os dados estão inconsistentes: alguns têm espaços extras, outros estão em letras minúsculas e há códigos duplicados que não podem ser cadastrados duas vezes no sistema.

lista_skus_brutos = [
    " iphone-13-preto ",
    "GALAXY-S21-PRATA",
    "iphone-13-preto",   
    "  macbook-m1-cinza ",
    "galaxy-s21-prata "   
]

2. Remova espaços em branco e converta tudo para Letras Maiúsculas.
3. Garanta que a lista final não contenha códigos repetidos.


**Ex2: Sistema de descontos**

Contexto:
Você está programando o caixa de uma papelaria. Para queimar o estoque, o dono da loja criou uma regra de desconto automática:

Produtos Acima de R$ 10,00: Ganham 10% de desconto.

Produtos Populares (Até R$ 10,00): Ganham 5% de desconto.

Lista a ser usada inicialmente:

precos_originais = [5.00, 15.00, 8.00, 20.00, 10.00]


O que o código deve fazer:

Percorrer a lista de preços originais.

Verificar o valor de cada item usando um if.

Aplicar o desconto correto e salvar o novo valor em uma lista chamada precos_com_desconto.


In [16]:
#Resolução ex1: Tratamento de dados de SKUs (produtos)

# Dados brutos vindos do fornecedor
lista_skus_brutos = [
    " iphone-13-preto ",
    "GALAXY-S21-PRATA",
    "iphone-13-preto",    # Duplicado
    "  macbook-m1-cinza ",
    "galaxy-s21-prata "   # Duplicado (em minúsculo)
]

# Lista para armazenar os dados tratados e únicos
lista_skus_limpos = []

for item in lista_skus_brutos:
    # 1. Tratamento: Remove espaços e coloca em MAIÚSCULO
    sku_tratado = item.strip().upper()

    # 2. Verificação de Duplicidade:
    # Só adicionamos se o SKU tratado ainda não estiver na lista final
    if sku_tratado not in lista_skus_limpos:
        lista_skus_limpos.append(sku_tratado)
        print(f"SKU Aceito: {sku_tratado}")
    else:
        print(f"SKU Duplicado Ignorado: {sku_tratado}")

print("\n--- Relatório Final de SKUs Únicos ---")
print(lista_skus_limpos)

SKU Aceito: IPHONE-13-PRETO
SKU Aceito: GALAXY-S21-PRATA
SKU Duplicado Ignorado: IPHONE-13-PRETO
SKU Aceito: MACBOOK-M1-CINZA
SKU Duplicado Ignorado: GALAXY-S21-PRATA

--- Relatório Final de SKUs Únicos ---
['IPHONE-13-PRETO', 'GALAXY-S21-PRATA', 'MACBOOK-M1-CINZA']


In [18]:
## Resolução ex2: Sistema de descontos
precos_originais = [5.00, 15.00, 8.00, 20.00, 10.00]

# Lista para guardar os novos valores calculados
precos_com_desconto = []

for preco in precos_originais:
    # Regra 1: Se for maior que 10 reais, desconto de 10% (multiplica por 0.90)
    if preco > 10:
        novo_valor = preco * 0.90

    # Regra 2: Caso contrário (menor ou igual a 10), desconto de 5% (multiplica por 0.95)
    else:
        novo_valor = preco * 0.95

    # Adiciona o resultado na lista final
    precos_com_desconto.append(novo_valor)

# Exibe as duas listas para comparação
print(f"Preços Originais: {precos_originais}")
print(f"Preços com Desconto: {precos_com_desconto}")

Preços Originais: [5.0, 15.0, 8.0, 20.0, 10.0]
Preços com Desconto: [4.75, 13.5, 7.6, 18.0, 9.5]
